src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)
+
+**ECMAScript specification types** are types that occur only in the ECMAScript spec, but not in the JavaScript language, and which might or might not occur as internal types in a JavaScript engine. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).
+
+[Spec: ECMAScript specification types](https://tc39.es/ecma262/#sec-ecmascript-specification-types)
+
+## Abstract operations
+
+**Abstract operations** are functions defined in the ECMAScript spec; they are defined for the purpose of writing the spec elegantly and concisely. A JavaScript engine doesn't have to implement them as separate functions inside the engine. They cannot be directly called from JavaScript.
+
+[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)
+
+## Internal slots and internal methods
+
+**Internal slots** and **internal methods** use names enclosed in \[\[ \]\].
+
+Internal slots are (assumed) data members (of a JavaScript object or a spec-internal data type) and internal methods are (assumed) member functions (of a JavaScript object or a spec-internal data type).
+
+Internal slots and methods are used in the algorithms described by the spec. Internal slots are used for storing the state of the object, and internal methods are functions associated with the object.
+
+[Spec: Object internal methods and internal slots](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots)
+
+For example, every JavaScript object has an internal slot \[\[Prototype\]\] and an internal method \[\[GetOwnProperty\]\].
+
+Internal slots and methods are not accessible from JavaScript, for example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.
+
+Each internal method delegates to a similarly-named abstract operation:
+
+> \[\[GetOwnProperty\]\](P)
+>
+> When the \[\[GetOwnProperty\]\] internal method of O is called with property key P, the following steps are taken:
+>
+> Return ! OrdinaryGetOwnProperty(O, P).
+
+[Spec: \[\[GetOwnProperty\]\]](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-getownproperty-p)
+
+(We'll find out what the exclamation mark means in the next chapter.)
+
+_OrdinaryGetOwnProperty_ is not an internal method, since it's not associated with any object; instead, the object it operates on is passed as a parameter.
+
+_OrdinaryGetOwnProperty_ is called "ordinary" since it operates on ordinary objects. EcmaScript objects can be either **ordinary** or **exotic**. Ordinary objects must have the default behavior for a set of methods called **essential internal methods**. If an object deviates from the default behavior, it's exotic.
```

Comment:
```suggestion
`OrdinaryGetOwnProperty` is called "ordinary" since it operates on ordinary objects. ECMAScript objects can be either _ordinary_ or _exotic_. Ordinary objects must have the default behavior for a set of methods called _essential internal methods_. If an object deviates from the default behavior, it's exotic.
```

Reply:
Did this suggestion *except* I kept the terms *bold* because they stand out better that way. (I'd like to be able to skim the blog post, look for a certain term and its explanation.)